syntax = "proto2";

package spiral.fluid;

message Config {
  message Brush {
    optional uint32 num_previous_speeds = 1 [default = 15];
    optional uint32 num_splats_per_segment = 2 [default = 8];

    optional uint32 num_bristles = 3 [default = 100];
    optional uint32 num_vertices_per_bristle = 4 [default = 10];
    optional float bristle_length = 5 [default = 4.5];
    optional float bristle_jitter = 6 [default = 0.5];

    optional uint32 num_iterations = 7 [default = 20];
    optional float gravity = 8 [default = 30.0];
    optional float brush_damping = 9 [default = 0.75];
    optional float stiffness_variation = 10 [default = 0.3];
  }

  message Simulator {
    optional uint32 num_jacobi_iterations = 1 [default = 2];
    optional uint32 num_frames_to_simulate = 2 [default = 60];
    optional float splat_padding = 3 [default = 4.5];
    optional float speed_padding = 4 [default = 1.1];
  }

  optional float brush_height = 1 [default = 2.0];
  optional float z_threshold = 2 [default = 0.13333];

  optional float splat_velocity_scale = 3 [default = 0.14];
  optional float splat_radius = 4 [default = 0.05];

  optional float min_alpha = 5 [default = 0.002];
  optional float max_alpha = 6 [default = 0.025];

  optional float normal_scale = 7 [default = 7.0];
  optional float roughness = 8 [default = 0.075];
  optional float f0 = 9 [default = 0.05];
  optional float specular_scale = 10 [default = 0.5];
  optional float diffuse_scale = 11 [default = 0.15];

  message Direction {
    optional float x = 1 [default = 0.0];
    optional float y = 2 [default = 1.0];
    optional float z = 3 [default = 1.0];
  }

  optional Direction light_direction = 12;

  optional Brush brush = 13;
  optional Simulator simulator = 14;
}
